crate/operator/utils/kopf.py

```diff
@@ -107,7 +140,51 @@ async def send_notifications(self, logger: logging.Logger):
                 notification["status"],
                 logger,
             )
-        self.clear_notifications()
-
-    def clear_notifications(self):
-        self._context.clear_notifications()
+        self._context.get("notifications", []).clear()
+
+    def _get_status(self, statuses: dict, dependency: str, logger) -> Optional[dict]:
+        """
+        Get the status of the specified dependency, obeying the ref.
+        """
+        status = statuses.get(dependency, None)
+        if not status:
+            return None
+
+        if status.get("ref", None) != self.ref:
+            logger.debug(
+                "Ignoring status for '%s' from previous run: %s", dependency, status
+            )
+            return None
+
+        return status
+
+    def _should_run_on_failed_dependency(
+        self, annotations: dict, handler_name: str, logger: logging.Logger
+    ) -> bool:
+        """
+        There is no way in kopf to say if a certain handler has failed or not.
+
+        What we are doing instead is peeking into kopf's internal state storage -
+        the annotations on the CrateDB objects to check if the handler has failed.
+
+        Slightly naughty, but there is no better way at the time of writing.
+        """
+        # Handler names have dots instead of slashes in annotations
+        normalized_name = handler_name.replace("/", ".")
+        key = f"{KOPF_STATE_STORE_PREFIX}/{normalized_name}"
+        status_str = annotations.get(key, None)
```

Comment:
`None` is the default of get
```suggestion
        status_str = annotations.get(key)
```